### ISSUE TITLE
CHANNEL_ERROR - second 'channel.open' seen

### ISSUE THREAD

**Issue by maxbrieiev** [squaremo/amqp.node#441]
`createChannel()` function starts throwing exceptions if it is called under certain circumstances.
Please, check following self-contained code sample.

- two consumers for a queue are being created;
- but only one consumer is allowed to consume from a given queue (by using `exclusive` flag)
- so an attempt to create a second consumer fails fails with `ACCESS_REFUSED` error. This is expected, but
- any subsequent attempt to call `connection.createChannel()` will fail with `CHANNEL_ERROR - second 'channel.open' seen`.

Environment:
- amqplib 0.5.2;
- node 10;
- rabbitmq server: 3.6.6 and 3.7.7

```javascript
'use strict';

const amqp = require('amqplib');
const url = 'amqp://guest:guest@localhost:5672';

function createConsumer(conn) {
  return conn
    .createChannel()
    .then(function (chan) {
      chan.on('error', () => null);
      return chan
        .assertQueue('queue_with_exclusive_consumer')
        .then(function (q) {
          return chan
            .consume(q.queue, console.log, {noAck: false, exclusive: true})
        });
    });
}

amqp
  .connect(url)
  .then(function (conn) {

    let firstConsumerP = createConsumer(conn).catch(console.log);

    let secondConsumerP = firstConsumerP
      .then(() => createConsumer(conn))
      .catch(
        e => e.message.includes('exclusive'),
        e => console.error(e.message)
      )
      .then(() => conn.createChannel()); // Boom!!!
  });
```
This outputs the following:
```
We are skipping this: Operation failed: BasicConsume; 403 (ACCESS-REFUSED) with message "ACCESS_REFUSED - queue 'queue_with_exclusive_consumer' in vhost '/' in exclusive use"
events.js:167
      throw er; // Unhandled 'error' event
      ^

Error: Connection closed: 504 (CHANNEL-ERROR) with message "CHANNEL_ERROR - second 'channel.open' seen"
    at Object.accept (/home/mbreev/dev/pinapp/lib/rmq/node_modules/amqplib/lib/connection.js:91:15)
    at Connection.mainAccept [as accept] (/home/mbreev/dev/pinapp/lib/rmq/node_modules/amqplib/lib/connection.js:64:33)
    at Socket.go (/home/mbreev/dev/pinapp/lib/rmq/node_modules/amqplib/lib/connection.js:477:48)
    at Socket.emit (events.js:182:13)
    at emitReadable_ (_stream_readable.js:531:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
Emitted 'error' event at:
    at Connection.emit (events.js:182:13)
    at Connection.C.onSocketError (/home/mbreev/dev/pinapp/lib/rmq/node_modules/amqplib/lib/connection.js:352:10)
    at Connection.emit (events.js:182:13)
    at Socket.go (/home/mbreev/dev/pinapp/lib/rmq/node_modules/amqplib/lib/connection.js:480:12)
    at Socket.emit (events.js:182:13)
    at emitReadable_ (_stream_readable.js:531:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)

```

This code worked under node 6. I am not sure about other versions. But it does fail under node 10.

Thank you

**Comment by KristjanTammekivi** [comment]
I had a similar issue where server closed a channel and I immediately opened a new one. Started working when I downgraded Rabbit from 3.7.7-management to 3.6.16-management.

**Comment by ubnt-michals** [comment]
I have the exact same issue with Rabbit 3.6.10. The problem looks like to be here:

https://github.com/squaremo/amqp.node/blob/master/lib/channel.js#L395-L412

Then the channel is immediately released to be free channel

https://github.com/squaremo/amqp.node/blob/master/lib/channel.js#L159-L165

If then it's allocated again by calling `createChannel()` - the connection will error with `CHANNEL_ERROR - second 'channel.open' seen`

**Comment by KristjanTammekivi** [comment]
I might be wrong in my interpretation but it seems like server is sending a frame claiming the channel is closed. I think amqp.node is right to flag that channel as free

**Comment by ubnt-michals** [comment]
Updated my original post. It only happens in case of error and yes channel is marked as free but not on the server! Unfortunately this renders amqplib quite unusable for us atm. I'll try to make a PR to fix this.

**Comment by squaremo** [comment]
NodeJS 10 changed the implementation of streams (the `'readable'` event is now always fired on the the next tick, rather than sometimes inline), and this invalidated an unwitting assumption in amqplib.

It seems to cause especial problems with the channel close protocol -- here, it looks like the CloseOK is not sent to the server before the channel ID is freed and a new channel opened. I think this can be fixed by making sure the channel buffer is ended before freeing the channel ID.

**Comment by bangbang93** [comment]
Any update with this?
I found that after upgrading node v10.11.0, I can reproduce 504 statable with following code
If the first channel is errored, the following channel absolutely error with 504

```typescript
import * as amqplib from 'amqplib'
import * as Bluebird from 'bluebird'

(async function () {
  const connection = await amqplib.connect('amqp://localhost')

  const p1 = () => connection.createChannel()
    .then(async (ch) => {
      console.log(1)
      ch.on('error', console.error)
      return ch.checkExchange('no.such.exchange.1')
    })
  const p2 = () => connection.createChannel()
    .then(async (ch) => {
      console.log(2)
      ch.on('error', console.error)
      return ch.checkExchange('no.such.exchange.2')
    })

  await p1().catch(() => {})
  await Bluebird.delay(1000)
  await p2().catch(() => {})
  await Bluebird.delay(1000)

  await connection.createChannel()
})()
```
![image](https://user-images.githubusercontent.com/3430784/47281818-60b30a80-d60f-11e8-9bcf-0a2f0a418b63.png)

nodejs v10.11.0
amqplib 0.5.2 and 0.5.1

**Comment by squaremo** [comment]
I believe this is fixed in master, and will shortly be in a release. Hold tight!

**Comment by MedusaLeee** [comment]
@squaremo Hi,I have same issue,so when will release 0.5.3? Thanks.

**Comment by Maxpain177** [comment]
Any updates?

**Comment by Maxpain177** [comment]
@squaremo

**Comment by bangbang93** [comment]
Is any plan here?

**Comment by KristjanTammekivi** [comment]
I tested out the fix from https://github.com/squaremo/amqp.node/commit/c973adbf3fc4a86758fb6e48a07a013ca4df327e and I can confirm it fixes my tests. If @squaremo could publish it to NPM I would really appreciate it!

**Comment by cressie176** [comment]
We released v0.5.3 today. Hopefully this should resolve the issue.
